File: src/cpu/probes/UserInstCounter.py
from m5.objects import SimObject
from m5.objects.Probe import ProbeListenerObject
from m5.params import *
from m5.util.pybind import *


class UserInstCounterManager(SimObject):
    type = "UserInstCounterManager"
    cxx_header = "cpu/probes/user_inst_counter.hh"
    cxx_class = "gem5::UserInstCounterManager"

    cxx_exports = [
        PyBindMethod("getUserInstCount"),
        PyBindMethod("getNoneUserInstCount"),
        PyBindMethod("resetUserInstCount"),
        PyBindMethod("resetNoneUserInstCount"),
    ]


class UserInstCounter(ProbeListenerObject):
    type = "UserInstCounter"
    cxx_header = "cpu/probes/user_inst_counter.hh"
    cxx_class = "gem5::UserInstCounter"

    if_start_listening = Param.Bool(True, "Start listening to instructions")
    counter_manager = Param.UserInstCounterManager(
        "The UserInstCounter manager"
    )

File: src/cpu/probes/user_inst_counter.cc
#include "cpu/probes/user_inst_counter.hh"

namespace gem5
{
UserInstCounterManager::UserInstCounterManager(
                                const UserInstCounterManagerParams &params) :
    SimObject(params),
    userInstCount(0),
    noneUserInstCount(0)
{
}

UserInstCounter::UserInstCounter(const UserInstCounterParams &params) :
    ProbeListenerObject(params),
    ifStartListening(params.if_start_listening),
    manager(params.counter_manager)
{
    if (params.counter_manager == nullptr){
        fatal("UserInstCounter must have a UserInstCounterManager\n");
    }
}

void
UserInstCounter::regProbeListeners()
{
    if (ifStartListening){
        typedef ProbeListenerArg<UserInstCounter, const bool>
                                                UserInstCounterProbeListener;
        listeners.push_back(new UserInstCounterProbeListener(this,
                            "CommitUserInst", &UserInstCounter::checkInst));
    }
}

void
UserInstCounter::startListening()
{
    if (listeners.size() == 0) {
        typedef ProbeListenerArg<UserInstCounter, const bool> U
                                                serInstCounterProbeListener;
        ifStartListening = true;
        listeners.push_back(new UserInstCounterProbeListener(this,
                            "CommitUserInst", &UserInstCounter::checkInst));
    }

}

void
UserInstCounter::stopListening()
{
    ifStartListening = false;
    listeners.clear();
}

void
UserInstCounter::checkInst(const bool &isUserInst)
{
    if (isUserInst){
        manager->increaseUserInstCount();
    } else {
        manager->increaseNoneUserInstCount();
    }
}

} // namespace gem5
